kernelci/storage/azure.py

```diff
@@ -0,0 +1,67 @@
+# SPDX-License-Identifier: LGPL-2.1-or-later
+#
+# Copyright (C) 2023 Collabora Limited
+# Author: Guillaume Tucker <guillaume.tucker@collabora.com>
+
+"""KernelCI storage implementation for Azure Files"""
+
+from urllib.parse import urljoin
+from azure.storage.fileshare import ShareServiceClient
+from . import Storage
+
+
+class StorageAzureFiles(Storage):
+    """Storage implementation for Azure Files
+
+    This class implements the Storage interface for uploading files to Azure
+    Files.  It uses a public Shared Access Signature read-only token for the
+    download URLs while also relying on a separate token with appropriate
+    permissions for uploading files passed as the storage credentials.
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._service = None
+
+    def _connect(self):
+        if self._service is not None:
+            return
+        self._service = ShareServiceClient(
+            account_url=self.config.base_url,
+            credential=self.credentials
+        )
+
+    def _get_share(self):
+        share = self._service.get_share_client(share=self.config.share)
+        share_found = None
+        for item in self._service.list_shares():
+            if item['name'] == self.config.share:
+                share_found = item
+                break
+        if share_found is None:
+            share.create_share()
+        return share
+
+    def _get_directory(self, share, path):
+        directory = share.get_directory_client(directory_path=path)
+        if not directory.exists():
```

Comment:
I think it is a little bit confusing logic, in _get_directory, if directory doesn't exist - create it.
IMO, it is better to handle that in _upload, if _get_directory returned None, create directory, then use it.
Otherwise, if for example in future this class will be extended, for example to check existence of particular results (directory), it wont be possible to use _get_directory for such purpose.

Reply:
This method is just an implementation detail for AzureFiles storage, the bass class method is just `upload()` and that's expected to create the destination path.  If we want to add methods for exploring storage we can do that later, at the moment it's not a supported feature.